client/client_test.go

```diff
@@ -7927,6 +7928,82 @@ func testMultipleCacheExports(t *testing.T, sb integration.Sandbox) {
 	ensureFileContents(t, filepath.Join(destDir, "unique"), string(uniqueFile))
 }
 
+// https://github.com/moby/buildkit/issues/3148
+func testMountStubsTimestamp(t *testing.T, sb integration.Sandbox) {
+	c, err := New(sb.Context(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	const sourceDateEpoch = int64(1234567890) // Fri Feb 13 11:31:30 PM UTC 2009
+	st := llb.Image("busybox:latest").Run(
+		llb.Args([]string{"/bin/touch", fmt.Sprintf("--date=@%d", sourceDateEpoch),
+			"/bin",
+			"/etc",
+			"/var",
+			"/var/foo",
+			"/tmp",
+			"/tmp/foo2",
+			"/tmp/foo2/bar",
+		}),
+		llb.AddMount("/var/foo", llb.Scratch(), llb.Tmpfs()),
+		llb.AddMount("/tmp/foo2/bar", llb.Scratch(), llb.Tmpfs()),
+	)
+	def, err := st.Marshal(sb.Context())
+	require.NoError(t, err)
+
+	tmpDir := t.TempDir()
+	tarFile := filepath.Join(tmpDir, "out.tar")
+	tarFileW, err := os.Create(tarFile)
+	require.NoError(t, err)
+	defer tarFileW.Close()
+
+	_, err = c.Solve(sb.Context(), def, SolveOpt{
+		Exports: []ExportEntry{
+			{
+				Type:   ExporterTar,
+				Output: fixedWriteCloser(tarFileW),
+			},
+		},
+	}, nil)
+	require.NoError(t, err)
+	tarFileW.Close()
+
+	tarFileR, err := os.Open(tarFile)
+	require.NoError(t, err)
+	defer tarFileR.Close()
+	tarR := tar.NewReader(tarFileR)
+	touched := map[string]*tar.Header{
+		"bin/": nil, // Regular dir
+		"etc/": nil, // Parent of file mounts (etc/{resolv.conf, hosts})
+		"var/": nil, // Parent of dir mount (var/foo)
+		"tmp/": nil, // Grandparent of dir mount (tmp/foo2/bar)
+		// No support for reproducing the timestamps of mount point directories such as /dev,
+		// because the touched timestamp value is lost when the mount is unmounted.
+	}
+	mustNotExist := map[string]struct{}{
+		"var/foo":  {}, // Created on mounting var/foo, and removed on unmounting it
+		"tmp/foo2": {}, // Created on mounting tmp/foo2/bar, and removed on unmounting it
```

Comment:
I'm a bit confused about what actually removes this parent dir of the mount. The described behavior is correct but I can't find the code that actually does that.

Reply:
Me too, shall we remove this check?

Reply:
The test turned out to be wrong. Directory entries must have the '/' suffix in their names.
Will update the test.

Reply:
done

Reply:
done